ambry-network/src/main/java/com.github.ambry.network/Selector.java

```diff
@@ -455,6 +481,9 @@ private void close(SelectionKey key) {
       activeConnections.set(this.keyMap.size());
       try {
         transmission.close();
+        if (pendingHandshakeTransmissions.get(transmission.getConnectionId()) != null) {
+          pendingHandshakeTransmissions.remove(transmission.getConnectionId());
+        }
       } catch (IOException e) {
```

Comment:
From the documentation it looks like the `Selector` isn't meant to be thread safe. The documentation of the class on line 64 says so. There are a lot of parts in this class that work on the fact that this class was not meant to be thread safe.